staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go

```diff
@@ -217,54 +300,144 @@ func (s *openAPIAggregator) handlerWithUser(handler http.Handler, info user.Info
+			if len(etag) > 0 && strings.HasPrefix(etag, locallyGeneratedEtagPrefix) {
+				// The function call with an etag and server does not report an etag.
+				// That means this server does not support etag and the etag that passed
+				// to the function generated previously by us. Just compare etags and
+				// return StatusNotModified if they are the same.
+				if etag == newEtag {
+					return nil, etag, http.StatusNotModified, nil
+				}
+			}
+		}
+		return openApiSpec, newEtag, http.StatusOK, nil
 	default:
-		return nil, fmt.Errorf("failed to retrive openAPI spec, http error: %s", writer.String())
+		return nil, "", 0, fmt.Errorf("failed to retrieve openAPI spec, http error: %s", writer.String())
 	}
 }
 
-// loadApiServiceSpec loads OpenAPI spec for the given API Service and then updates aggregator's spec.
-func (s *openAPIAggregator) loadApiServiceSpec(handler http.Handler, apiService *apiregistration.APIService) error {
+// AddApiServiceSpec add the api service to OpenAPI controller queue to be loaded.  It is thread safe.
+func (s *openAPIAggregator) AddApiServiceSpec(handler http.Handler, apiService *apiregistration.APIService) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
 
 	// Ignore local services
 	if apiService.Spec.Service == nil {
-		return nil
+		return
 	}
 
-	openApiSpec, err := s.downloadOpenAPISpec(handler)
-	if err != nil {
-		return err
+	if _, existingService := s.openAPISpecs[apiService.Name]; existingService {
+		s.openAPIAggregationController.UpdateAPIService(apiService.Name)
+		return
 	}
-	aggregator.FilterSpecByPaths(openApiSpec, []string{"/apis/" + apiService.Spec.Group + "/"})
 
 	s.openAPISpecs[apiService.Name] = &openAPISpecInfo{
 		apiService: *apiService,
-		spec:       openApiSpec,
+		spec:       nil,
+		handler:    handler,
+		etag:       "",
 	}
+	s.openAPIAggregationController.AddAPIService(apiService.Name)
+}
+
+// RemoveApiServiceSpec removes an api service from OpenAPI aggregation. It is thread safe.
+func (s *openAPIAggregator) RemoveApiServiceSpec(apiServiceName string) error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
 
-	err = s.updateOpenAPISpec()
+	return s.removeApiServiceSpec(apiServiceName)
+}
+
+// removeApiServiceSpec is the internal version of RemoveApiServiceSpec that does not hold any lock and safe to call
+// from a method that holds the lock.
+func (s *openAPIAggregator) removeApiServiceSpec(apiServiceName string) error {
+	if _, existingService := s.openAPISpecs[apiServiceName]; !existingService {
+		return nil
+	}
+	oldSpecs := s.openAPISpecs
+	delete(s.openAPISpecs, apiServiceName)
+	err := s.updateOpenAPISpec()
 	if err != nil {
-		delete(s.openAPISpecs, apiService.Name)
+		s.openAPISpecs = oldSpecs
 		return err
 	}
 	return nil
 }
+
+// UpdateApiServiceSpec updates and aggregation OpenAPI spec for given service name. It is thread safe.
+func (s *openAPIAggregator) UpdateApiServiceSpec(apiServiceName string) (shouldBeRateLimited, deleted bool, err error) {
+	s.mutex.Lock()
```

Comment:
this lock is hold during the whole download (with 60s timeout)? So Add/RemoveApiServiceSpec are still blocked by the download.